File: Practice Q/Day2/Java/balanced_parentheses.java
import java.util.Stack;

public class Main {
    public static boolean isBalanced(String expr) {
        Stack<Character> s = new Stack<>();

        for (char ch : expr.toCharArray()) {
            if (ch == '(')
                s.push(ch);
            else if (ch == ')') {
                if (s.isEmpty())
                    return false;
                s.pop();
            }
        }

        return s.isEmpty();
    }

    public static void main(String[] args) {
        String expr = "((()))";
        System.out.println(isBalanced(expr) ? "Balanced" : "Not Balanced");
    }
}

File: Practice Q/Day2/Python/balanced_parentheses.py
def is_balanced(expr):
    stack = []

    for ch in expr:
        if ch == '(':
            stack.append(ch)
        elif ch == ')':
            if not stack:
                return False
            stack.pop()

    return len(stack) == 0

expr = "((()))"
print("Balanced" if is_balanced(expr) else "Not Balanced")
